pkg/pool-weighted/contracts/managed/ManagedPool.sol

```diff
@@ -80,11 +87,44 @@ contract ManagedPool is ManagedPoolSettings {
         _require(ManagedPoolStorageLib.getSwapsEnabled(poolState), Errors.SWAPS_DISABLED);
 
         uint256 weightChangeProgress = ManagedPoolStorageLib.getGradualWeightChangeProgress(poolState);
-        uint256 tokenInWeight = _getNormalizedWeight(request.tokenIn, weightChangeProgress);
-        uint256 tokenOutWeight = _getNormalizedWeight(request.tokenOut, weightChangeProgress);
+        uint256 swapFeePercentage = ManagedPoolStorageLib.getSwapFeePercentage(poolState);
 
-        uint256 swapFeeComplement = ManagedPoolStorageLib.getSwapFeePercentage(poolState).complement();
+        if (request.tokenOut == IERC20(this)) {
+            // Do a joinSwap
+        } else if (request.tokenIn == IERC20(this)) {
+            // Do an exitSwap
+        } else {
+            uint256 tokenInWeight = _getNormalizedWeight(request.tokenIn, weightChangeProgress);
+            uint256 tokenOutWeight = _getNormalizedWeight(request.tokenOut, weightChangeProgress);
+            return
+                _onTokenSwap(
+                    request,
+                    balanceTokenIn,
+                    balanceTokenOut,
+                    tokenInWeight,
+                    tokenOutWeight,
+                    swapFeePercentage.complement()
```

Comment:
We'll need to read the scaling factors inside this funcion, which will read the same state `_getNormalizedWeight` has already accesses. It also doesn't seem like we gain much by computing the weights outside of it.

What if this function simply read the pool state, checked for swaps to be enabled, and then dispatched to all three handlers, passing the state and leaving computation of fees, weights, etc. up to them?